sonic-ledd/scripts/ledd

```diff
@@ -0,0 +1,224 @@
+
+    # If we can't locate a platform-specific module, exit gracefully, assuming this
+    # platform utilizes a hardware-based LED control solution
+    if not os.path.isfile(module_file):
+        log_info("Failed to locate platform-specific %s module. Exiting..." % LED_MODULE_NAME)
+        sys.exit(0)
+
+    try:
+        module = imp.load_source(LED_MODULE_NAME, module_file)
+    except IOError, e:
+        log_error("Failed to load platform module '%s': %s" % (LED_MODULE_NAME, str(e)))
+        return -1
+
+    log_info("Loaded module '%s'." % LED_MODULE_NAME)
+
+    try:
+        led_control_class = getattr(module, LED_CLASS_NAME)
+        led_control = led_control_class()
+    except AttributeError, e:
+        log_error("Failed to instantiate '%s' class: %s" % (LED_CLASS_NAME, str(e)))
+        return -2
+
+    log_info("Instantiated class '%s.%s'." % (LED_MODULE_NAME, LED_CLASS_NAME))
+
+    return 0
+
+#=============================== Main ================================
+
+def main():
+    port_status_dict = {}
+
+    log_info("Starting up...")
+
+    if not os.geteuid() == 0:
+        log_error("Must be root to run this daemon")
+        print "Error: Must be root to run this daemon"
+        sys.exit(1)
+
+    # Parse options if provided
+    if (len(sys.argv) > 1):
+        try:
+            options, remainder = getopt.getopt(sys.argv[1:],
+                    'hv',
+                    ['help',
+                     'version'])
+        except getopt.GetoptError, e:
+            print e
+            print USAGE_HELP
+            sys.exit(2)
+
+        for opt, arg in options:
+            if opt == '--help' or opt == '-h':
+                print USAGE_HELP
+                sys.exit(0)
+            elif opt == '--version' or opt == '-v':
+                print 'ledd version ' + VERSION
+                sys.exit(0)
+
+    # Register our signal handlers
+    signal.signal(signal.SIGHUP, signal_handler)
+    signal.signal(signal.SIGINT, signal_handler)
+    signal.signal(signal.SIGTERM, signal_handler)
+
+    # Load platform-specific LedControl class
+    err = load_platform_led_control_module()
+    if err != 0:
+        sys.exit(1)
+
+    # Connect to APPL_DB using SwSS SDK
+    swss = swsssdk.SonicV2Connector()
+    swss.connect(swss.APPL_DB)
+
+    # Get a list of all ports from the database
+    port_table_keys = swss.keys(swss.APPL_DB, PORT_TABLE_PREFIX + '*');
+
+    # Create a dictionary of <sonic_port_name>:<link_status> containing all ports
+    # Initially set all statuses to 'down'
+    for key in port_table_keys:
+        # Remove table name prefix
+        port_name = key[len(PORT_TABLE_PREFIX):]
+
+        # TODO: Once the DB is fixed and this 'ConfigDone' entry is gone, remove this check
+        if port_name == 'ConfigDone':
+            continue
+
+        port_status_dict[port_name] = 'down'
+        led_control.port_link_state_change(port_name, port_status_dict[port_name])

```

Comment:
one problem is that you only get the port_table once, it may not be a complete list.